Implement metagroup handling for apods

swagner-de opened the pull request.

swagner-de committed:
Enable threading so paginated results can be fetched in parallel

swagner-de committed:
Implement metagroup handling for apods

sebageek commented on networking_ccloud/tools/netbox_config_gen.py:
```diff
@@ -481,40 +491,49 @@ def make_vpod_metagroup(self, cluster: NbRecord, member: NbRecord) -> Optional[c
+        m = self.apod_node_name_re.match(member.name)
+        if not m:
+            raise ValueError(f'{member.name} is not complying with expected node naming for apod clusters')
+        apod_sequence = int(m.group(1))
```
If you'd name your catchgroups in the regex you could use `m.group('node')` (or apod) here and it would a bit more clear what's going on :)

sebageek commented on networking_ccloud/tools/netbox_config_gen.py:
```diff
@@ -481,40 +491,49 @@ def make_vpod_metagroup(self, cluster: NbRecord, member: NbRecord) -> Optional[c
+        apod_sequence = int(m.group(1))
+        binding_host_prefix = f'neutron-network-agent-a{apod_sequence:03d}-'
+        # right now 15 network agents are statically created per apod
+        binding_hosts = [f'{binding_host_prefix}{x}' for x in range(0, 15)]
```
Maybe make this a constant? Could even be something configurable.

sebageek commented on networking_ccloud/tools/netbox_config_gen.py:
```diff
@@ -481,40 +491,49 @@ def make_vpod_metagroup(self, cluster: NbRecord, member: NbRecord) -> Optional[c
-        # Check cluster names for duplicates
-        cluster_names = (x.name for x in metagroups.keys())
-        for name, count in Counter(cluster_names).items():
+        # Check cluster names for duplicates5
```
`duplicates5` - the new fancy check in town! ;)

sebageek commented on networking_ccloud/tools/netbox_config_gen.py:
```diff
@@ -481,40 +491,49 @@ def make_vpod_metagroup(self, cluster: NbRecord, member: NbRecord) -> Optional[c
+            # unfortunately the assumption that a netbox cluster corresponds to a metagroup does not hold with
+            # with apods anymore, hence, we call the metagroup handler and then use the binding_host to check
+            # if that metagroup exists
+            binding_hosts = tuple(metagroup.binding_hosts)
```
You could also use the property `binding_host_name` of the `Hostgroup` as key if you want:
```
    @property
    def binding_host_name(self):
        """Generate a name for this hostgroup by joining all binding hosts"""
        return ",".join(self.binding_hosts)
```

swagner-de committed:
Allow interface naming for breakoutable interfaces or such that are not

We currenlty are not using the interface slot or base interface property anyhow
so we can make it optional

swagner-de committed:
Fixes for dry-run mode

When running in dry-run mode, objects might or might not exist in netbox. If
they don't we use the dict that we'd elsewise use to create the object as a
mock. Unfortunately dicts are not hashable and also attribute access needs to
be treated a bit different. We address 2 occurence of these in this commit.

swagner-de committed:
Teach netbox modeller to handle apods

sebageek commented on networking_ccloud/tools/netbox_config_gen.py:
```diff
@@ -458,63 +502,42 @@ def get_interconnect_hostgroups(self, nb_switches: List[NbRecord]) -> List[conf.
 
         # Check cluster names for duplicates
-        cluster_names = (x.name for x in metagroups.keys())
-        for name, count in Counter(cluster_names).items():
+        for name, count in Counter(metagroups.keys()).items():
             if count > 1:
                 raise ConfigException(f'Cluster {name} appeared {count} times')
```
I don't think this check will ever work, as you are checking if there are multiple entries with the same key in a dictionary. If you'd try that, you'd just override the old entry.

swagner-de replied:
Indeed, it did not do anything that made sense earlier and it does not make sense now.

swagner-de committed:
Remove check for duplicates, it's already a dict and has no duplicates
